### DISCUSSION TITLE
How to forward all go2rtc/webrtc streams to Home Assistant?

### DISCUSSION THREAD

**Discussion by stalakerob** [blakeblackshear/frigate#11992] [Ask A Question] [answered]
### What is your question:

I have up to 3 streams per camera and the frigate custom component configured in HA. It seems only the main stream is available as a camera entity in HA or my config is missing something? For example I'm only getting camera.gartencam but not camera.gartencam_sub. 

`mqtt:
  enabled: true
  host: 192.168.1.200
  port: 1883
  user: user	
  password: password

detectors:
  coral:
    type: edgetpu
    device: usb

objects:
  track:
    - person
    - dog
    - cat
    - bird
    - car

snapshots:
  enabled: true
  retain:
    default: 7
record:
  enabled: true
#  sync_recordings: true
  retain:
    days: 7
    mode: all
  events:
    pre_capture: 3
    post_capture: 3
    retain:
      default: 7
      mode: motion

birdseye:
  # Optional: Enable birdseye view (default: shown below)
  enabled: true
  # Optional: Restream birdseye via RTSP (default: shown below)
  # NOTE: Enabling this will set birdseye to run 24/7 which may increase CPU usage somewhat.
  mode: continuous

go2rtc:
  streams:
    GartenCam:
      - rtsp://admin:xxxxxxx@192.168.1.203:554/h264Preview_01_main
    GartenCam_sub:
      - rtsp://admin:xxxxxxx@192.168.1.203:554/h264Preview_01_sub
    GartenCam_ext:
      - http://192.168.1.205/flv?port=1935&app=bcs&stream=channel0_ext.bcs&user=admin&password=xxxxxx

    CarportCam:
      - rtsp://admin:xxxxxx@192.168.1.204:554/h264Preview_01_main
    CarportCam_sub:
      - rtsp://admin:xxxxxx@192.168.1.204:554/h264Preview_01_sub
    CarportCam_ext:
      - http://192.168.1.204/flv?port=1935&app=bcs&stream=channel0_ext.bcs&user=admin&password=xxxxxx

    WZCam:
      - rtsp://admin:xxxxxx@192.168.1.206:554/h264Preview_01_main
    WZCam_sub:
      - rtsp://admin:xxxxxx@192.168.1.206:554/h264Preview_01_sub

    GartenCam2:
      - rtsp://admin:xxxxxx@192.168.1.208:554/h264Preview_01_main
    GartenCam2_sub:
      - rtsp://admin:xxxxxx@192.168.1.208:554/h264Preview_01_sub
    GartenCam2_ext:
      - http://192.168.1.208/flv?port=1935&app=bcs&stream=channel0_ext.bcs&user=admin&password=xxxxxx

  webrtc:
    candidates:
      - 192.168.1.244:8555
      - stun:8555

cameras:
  GartenCam: # <------ Name the camera
    enabled: true
    motion:
      threshold: 15
      contour_area: 8
      improve_contrast: 'true'
      mask:
        - 0.708,0.941,1,0.835,1,1,0.709,1
        - 0.841,0.139,0.835,0.618,1,0.686,1,0.15
    ffmpeg:
#      hwaccel_args: preset-intel-qsv-h264
      inputs:
        - path: rtsp://127.0.0.1:8554/GartenCam # <--- the name here must match the name of the camera in restream
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/GartenCam # <----- The stream you want to use for detection
          input_args: preset-rtsp-restream
          roles:
            - detect
    objects:
      track:
        - person
        - dog
        - cat
        - bird
    detect:
      enabled: true # <---- disable detection until you have a working camera feed
#      width: 640
#      height: 360
#      fps: 7

    live:
      stream_name: GartenCam_ext

  GartenCam2: # <------ Name the camera
    enabled: true
    motion:
      mask: 0.695,0.948,0.699,0.981,1,0.984,0.995,0.945
      threshold: 15
      contour_area: 10
      improve_contrast: 'true'
    ffmpeg:
#      hwaccel_args: preset-intel-qsv-h264
      inputs:
        - path: rtsp://127.0.0.1:8554/GartenCam2 # <--- the name here must match the name of the camera in restream
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/GartenCam2_sub # <----- The stream you want to use for detection
          input_args: preset-rtsp-restream
          roles:
            - detect
    objects:
      track:
        - person
        - dog
        - cat
        - bird
    detect:
      enabled: true # <---- disable detection until you have a working camera feed
    live:
      stream_name: GartenCam2_ext

  CarportCam: # <------ Name the camera
    enabled: true
    motion:
      mask:
        - 640,455,640,480,444,480,445,452
      threshold: 20
      contour_area: 10
      improve_contrast: 'true'
    ffmpeg:
#      hwaccel_args: preset-intel-qsv-h264
      inputs:
        - path: rtsp://127.0.0.1:8554/CarportCam # <--- the name here must match the name of the camera in restream
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/CarportCam_sub # <----- The stream you want to use for detection
          input_args: preset-rtsp-restream
          roles:
            - detect
    objects:
      track:
        - person
        - dog
        - cat
        - car
    detect:
      enabled: true # <---- disable detection until you have a working camera feed
    live:
      stream_name: CarportCam_ext

  WZCam: # <------ Name the camera
    enabled: true
    motion:
      mask:
        - 640,322,438,319,441,352,640,352,640,352
    ffmpeg:
#      hwaccel_args: preset-intel-qsv-h264
      inputs:
        - path: rtsp://127.0.0.1:8554/WZCam # <--- the name here must match the name of the camera in restream
          input_args: preset-rtsp-restream
          roles:
            - record
        - path: rtsp://127.0.0.1:8554/WZCam_sub # <----- The stream you want to use for detection
          input_args: preset-rtsp-restream
          roles:
            - detect
    objects:
      track:
        - person
    detect:
      enabled: true # <---- disable detection until you have a working camera feed
    live:
      stream_name: WZCam_sub

camera_groups:
  My_Group:
    order: 1
    icon: LuVideo
    cameras:
      - CarportCam
      - GartenCam
      - GartenCam2
      - WZCam
  Test:
    order: 2
    icon: LuWebcam
    cameras:
      - GartenCam
      - GartenCam2
version: 0.14
`

**Comment by NickM-27** [comment]
This is working as expected. The frigate card in HA supports using more streams but there is only one camera entity.

**Reply by stalakerob** [reply]
Thanks for you quick answer. I tried adding streams to the frigate hass card but it did not work. I don't know how to exactly add a frigate substream to it. Also with VLC the substream is not available, at least not using rtsp://192.168.1.244:8554/gartencam_sub. I tried this:

`type: custom:frigate-card
cameras:
  - camera_entity: camera.gartencam
    live_provider: go2rtc
    dependencies:
      cameras:
        - gartencam_sub
  - camera_entity: camera.gartencam
    live_provider: go2rtc
    id: gartencam_sub
    hide: true
    go2rtc:
      modes: []
      stream: rtsp://192.168.1.244:8554/gartencam_sub

**Reply by ofcedwards** [reply]
You likely have some other error in your configuration that's blocking this. Have you tried using the rtsp url of your main stream to see if that works? For me, any stream that's configured under ````go2rtc```` in the frigate config is available at ````rtsp://FRIGATE_URL:8554/GO2RTC_STREAM_NAME````. This is true for me whether I have the stream set up to show in frigate or not. Simply having it under the ````go2rtc```` section in the config makes it available.

**Reply by ofcedwards** [reply, marked as answer]
Also note that the stream names appear to be case-sensitive. You have capital letters in the stream names in your config, but not in the url you provided to pull the stream into VLC/HA.

**Reply by ofcedwards** [reply]
Specifically for what you're doing with the frigate card, I've tested this and it works:
````
type: custom:frigate-card
cameras:
  - frigate:
      url: http://FRIGATE_URL:5000
    dependencies:
      cameras:
        - front-door-low-res
    camera_entity: camera.front_door
    live_provider: go2rtc
  - camera_entity: camera.front_door #camera entity from frigate integration
    live_provider: go2rtc
    id: front-door-low-res
    hide: true
    go2rtc:
      stream: front_door_sub #case-sensitive stream name from frigate go2rtc config
````

**Reply by stalakerob** [reply]
Thanks for your support. I removed the webrtc configuration from my config file. Now the sub streams work with VLC. Does this make sense?

**Reply by ofcedwards** [reply]
I'm not sure that removing the ````webrtc```` config would make any difference or that it's recommended. I have webrtc configured with
````
webrtc:
  candidates:
    - FRIGATE_URL:8555
    - stun:8555
````
This works as expected and allows me to view the streams with webrtc. Is the only change you made removing that block?

**Reply by stalakerob** [reply]
Yes. BTW I forgot to mention that I'm using Frigate 0.14 Beta 3. Maybe this is the problem?

**Reply by stalakerob** [reply]
> Also note that the stream names appear to be case-sensitive
This was my main issue. Many thanks. Now it works. Also with webrtc enabled.
